core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -1849,30 +1850,26 @@ public void delete(AlluxioURI path, DeleteContext context)
               .getMessage(path));
         }
 
+        List<String> failedChildren = new ArrayList<>();
         if (context.getOptions().getRecursive()) {
-          List<String> failedChildren = new ArrayList<>();
-          try (LockedInodePathList descendants = mInodeTree.getDescendants(inodePath)) {
-            for (LockedInodePath childPath : descendants) {
-              try {
-                mPermissionChecker.checkPermission(Mode.Bits.WRITE, childPath);
-                if (mMountTable.isMountPoint(childPath.getUri())) {
-                  mMountTable.checkUnderWritableMountPoint(childPath.getUri());
-                }
-              } catch (AccessControlException e) {
-                failedChildren.add(e.getMessage());
+          List<MountInfo> childrenMountPoints = mMountTable.findChildrenMountPoints(path, true);
+          if (!childrenMountPoints.isEmpty()) {
+            // TODO(jiacheng): a better string?
+            LOG.debug("Deleting {} which contains mount points {}", path, childrenMountPoints);
+            for (MountInfo mount : childrenMountPoints) {
```

Comment:
If this path contains readonly mount points, the delete will fail before attempting to delete inodes. This might not align with the best-effort approach right now. Should this be changed to best-effort deleting instead? The code logic might get messier. 

Another approach is to disallow deleting a parent of multiple mount points. Instead, return a message that the user should manually delete mount points one by one. This will make the control much easier.

One thing I'm missing is, I'm not checking if the operation is under a writable mount point itself. I should address this.

Reply:
I think it's OK to start with this but let's raise a ticket to PM.